enhancements/installer/byo-custom-dns.md

```diff
@@ -0,0 +1,571 @@
+
+When it is indicated to the OpenShift installer via an input parameter that the
+custom DNS solution is in use, OpenShift does not configure the cloud DNS
+solution [Route53 for AWS](https://aws.amazon.com/route53/) for any of its
+services. In the case of API and API-Int services, the customer is responsible
+for creating the external and internal LBs [Elastic Load Balancer](https://aws.amazon.com/elasticloadbalancing/)
+prior to cluster installation. Their custom DNS solution also needs to be
+configured with entries for API and API-Int before cluster installation. OpenShift
+will continue to configure and manage these cloud LBs created by the user.
+
+DNS configuration for Ingress services will be handled differently. The Ingress
+Operator utilizes the upstream k8s LoadBalancer service to create and configure
+the LBs for the *.apps service. In this case, the customer pre-creating the LB
+would mean the LB service would have to be modified to consume a pre-created LB.
+Accomplishing this would require a change to the upstream LB service potentially
+adding delays towards the completion of this feature. Since the LB DNS Name for
+the *.apps would not be available before cluster installation, the DNS solution
+cannot be pre-configured before cluster installation either.
+
+To work around this limitation, the Ingress Operator would continue using the
+LoadBalancer service to create and configure the LBs for the *.apps service.
+OpenShift will start an in-cluster DNS pod to provide DNS resolution for *.apps
+during cluster installation. After the cluster is installed, the user will have
+to update their custom DNS solution with an entry for the *.apps service.
+
+
+### Workflow Description
+
+1. Customer creates an external cloud LB for the API service and an internal
+cloud LB for the API-Int service. Since these LBs need to be created on subnets,
+these subnets also need to be created before cluster installation. As a result,
+the installer created subnets can not longer be supported with an external DNS.
+These subnets are created within VPCs and so the customer has to create the
+VPCs too before creating the subnets. In the Managed Services (MS) environment,
+ the internal(?) LBs for rh-api also need to be pre-created by the customer.
+
+2. The customer then uses the LB DNS names generated after their creation to
+configure their custom DNS solution with entries for API, API-Int and if
+applicable rh-api (for MS).
+
+3. The customer then updates the install-config to turn on the custom-dns
+feature. They also provide the LB Names and types for the pre-created API and
+API-Int LBs. The Installer would use these LB Names to configure them.
+
+4. The Installer skips configuring the cloud DNS solution when this feature is
+enabled via install-config. Since API and API-Int resolution are pre-requisties
+for successful cluster installation, the Installer performs additional
+validations to make sure they are configured correctly.
```

Comment:
Good to have this check but is there any option to opt-out of this? At least at one point in time we said that the Installer did not require the same DNS view of the world as the running cluster and therefore this would need to be something which can be skipped.

Reply:
Not aware of the details around why the DNS view of the Installer can be different from a running cluster. A check was added sometime back tocheck for API and API-Int reachability in bootkube. If unreachable, we display a warning and continue to proceed with the Install. 
It appears that that should be sufficient and no additional strict validation is required.

Reply:
Scott mentioned this to me yesterday, but it took me a while to catch on. I think the two contrasting considerations are:

1. We validate as much as possible to prevent installs from starting when we know they will fail. 
2. It's possible that the cluster can resolve DNS records that the installer can't.  So the install can proceed to complete, but the installer will consider the install failed after it fails to connect to the API.

The point here is that the installer can't be certain that failed DNS resolution (from the installer) will predict a cluster failure, so previously we have decided to let installs go forward in those cases. 

> Not aware of the details around why the DNS view of the Installer can be different from a running cluster. 

this is possible because the installer and cluster can be (and most often are) on different networks.

Reply:
>  so previously we have decided to let installs go forward in those cases.

In those cases, we weren't depending on users to setup a piece of DNS infrastructure. The expectations may be different depending on whether it's a publish: Internal vs publish: External cluster. 

On a `publish: External` cluster it may be reasonable to require that the installer can resolve dns

When determining about this validation we probably want to gauge how concerned we are about failures due to customers providing infrastructure. If customers have correctly setup their DNS records, in theory we should be able to query the API url and check that the returned hostname corresponds to the load balancer provided in the install config. I'm not sure how challenging that would be in practice.

Reply:
Updated the enhancement to capture this detail.